File: cli/src/routing.rs
use url::Url;

use crate::routing_memory::RoutingMemory;

#[derive(Debug, Clone)]
pub struct PreflightResult {
    pub platform: Option<&'static str>,
    pub preferred_strategy: &'static str,
    pub confidence: f32,
    pub js_heavy: bool,
}

#[derive(Debug, Clone)]
pub struct ResolutionBudget {
    pub max_provider_attempts: usize,
    pub max_paid_attempts: usize,
    pub max_total_latency_ms: u64,
    pub allow_paid: bool,
    pub attempts: usize,
    pub paid_attempts: usize,
    pub elapsed_ms: u64,
    pub stop_reason: Option<String>,
}

impl ResolutionBudget {
    pub fn can_try(&mut self, is_paid: bool) -> bool {
        if self.attempts >= self.max_provider_attempts {
            self.stop_reason = Some("max_provider_attempts".into());
            return false;
        }
        if is_paid && !self.allow_paid {
            self.stop_reason = Some("paid_disabled".into());
            return false;
        }
        if is_paid && self.paid_attempts >= self.max_paid_attempts {
            self.stop_reason = Some("max_paid_attempts".into());
            return false;
        }
        if self.elapsed_ms >= self.max_total_latency_ms {
            self.stop_reason = Some("max_total_latency_ms".into());
            return false;
        }
        true
    }

    pub fn record_attempt(&mut self, is_paid: bool, latency_ms: u64) {
        self.attempts += 1;
        self.elapsed_ms += latency_ms;
        if is_paid {
            self.paid_attempts += 1;
        }
    }
}

#[derive(Debug, Clone)]
pub struct PlannedProvider {
    pub name: String,
    pub is_paid: bool,
    pub skip_reason: Option<String>,
}

pub fn detect_doc_platform(url: &str) -> Option<&'static str> {
    let parsed = Url::parse(url).ok()?;
    let hostname = parsed.host_str()?.to_ascii_lowercase();
    let path = parsed.path().to_ascii_lowercase();

    if hostname == "gitbook.io" || hostname.ends_with(".gitbook.io") {
        return Some("gitbook");
    }
    if hostname == "gitbook.com" || hostname.ends_with(".gitbook.com") {
        return Some("gitbook");
    }
    if hostname == "readthedocs.io" || hostname.ends_with(".readthedocs.io") {
        return Some("sphinx");
    }
    if hostname == "rtfd.io" || hostname.ends_with(".rtfd.io") {
        return Some("sphinx");
    }
    if hostname == "mkdocs.org" || hostname == "www.mkdocs.org" {
        return Some("mkdocs");
    }
    if hostname == "notion.so" || hostname.ends_with(".notion.so") {
        return Some("notion");
    }
    if hostname == "notion.site" || hostname.ends_with(".notion.site") {
        return Some("notion");
    }
    if (hostname.ends_with(".atlassian.net") && path.starts_with("/wiki"))
        || hostname.contains("confluence")
        || path.contains("confluence")
    {
        return Some("confluence");
    }

    None
}

pub fn preflight_route(url: &str) -> PreflightResult {
    let platform = detect_doc_platform(url);
    let parsed = Url::parse(url).ok();
    let hostname = parsed
        .as_ref()
        .and_then(|u| u.host_str())
        .map(str::to_ascii_lowercase)
        .unwrap_or_default();
    let path = parsed
        .as_ref()
        .map(|u| u.path().to_ascii_lowercase())
        .unwrap_or_default();

    if matches!(platform, Some("gitbook" | "sphinx" | "mkdocs")) {
        return PreflightResult {
            platform,
            preferred_strategy: "llms_txt",
            confidence: 0.85,
            js_heavy: false,
        };
    }

    if matches!(platform, Some("notion" | "confluence")) {
        return PreflightResult {
            platform,
            preferred_strategy: "extraction",
            confidence: 0.8,
            js_heavy: true,
        };
    }

    let doc_signals = [
        "docs.",
        "doc.",
        "documentation",
        "/docs/",
        "/doc/",
        "/api/",
        "/reference/",
    ];
    if doc_signals
        .iter()
        .any(|s| hostname.contains(s) || path.contains(s))
    {
        return PreflightResult {
            platform: None,
            preferred_strategy: "llms_txt",
            confidence: 0.6,
            js_heavy: false,
        };
    }

    if ["github.com", "gitlab.com", "bitbucket.org"]
        .iter()
        .any(|d| hostname.contains(d))
    {
        return PreflightResult {
            platform: None,
            preferred_strategy: "direct_fetch",
            confidence: 0.7,
            js_heavy: false,
        };
    }

    PreflightResult {
        platform: None,
        preferred_strategy: "llms_txt",
        confidence: 0.4,
        js_heavy: false,
    }
}

#[allow(clippy::too_many_arguments)]
pub fn plan_provider_order(
    target: &str,
    is_url: bool,
    custom_order: Option<&[String]>,
    skip_providers: &[String],
    routing_memory: Option<&RoutingMemory>,
) -> Vec<PlannedProvider> {
    let mut base: Vec<String> = if let Some(custom) = custom_order {
        custom
            .iter()
            .filter(|p| {
                if let Ok(pt) = p.parse::<crate::types::ProviderType>() {
                    if is_url {
                        pt.is_url_provider()
                    } else {
                        pt.is_query_provider()
                    }
                } else {
                    false
                }
            })
            .cloned()
            .collect()
    } else if is_url {
        let preflight = preflight_route(target);

        if matches!(preflight.platform, Some("notion" | "confluence")) || preflight.js_heavy {
            vec![
                "firecrawl".into(),
                "mistral_browser".into(),
                "jina".into(),
                "direct_fetch".into(),
                "duckduckgo".into(),
            ]
        } else if preflight.preferred_strategy == "direct_fetch" {
            vec![
                "direct_fetch".into(),
                "llms_txt".into(),
                "jina".into(),
                "firecrawl".into(),
                "mistral_browser".into(),
                "duckduckgo".into(),
            ]
        } else {
            vec![
                "llms_txt".into(),
                "jina".into(),
                "firecrawl".into(),
                "direct_fetch".into(),
                "mistral_browser".into(),
                "duckduckgo".into(),
            ]
        }
    } else {
        // DuckDuckGo deprioritized due to instability (Alert 2026-04-20)
        vec![
            "exa_mcp".into(),
            "exa".into(),
            "tavily".into(),
            "serper".into(),
            "mistral_websearch".into(),
            "duckduckgo".into(),
        ]
    };

    if let Some(memory) = routing_memory {
        let domain = if is_url {
            crate::resolver::cascade::extract_domain_or_default(target)
        } else {
            "query".to_string()
        };
        base = memory.rank_providers(&domain, &base);
    }

    base.into_iter()
        .filter(|p| !skip_providers.contains(p))
        .map(|name| PlannedProvider {
            is_paid: matches!(
                name.as_str(),
                "exa" | "tavily" | "firecrawl" | "mistral_browser" | "mistral_websearch" | "serper"
            ),
            name,
            skip_reason: None,
        })
        .collect()
}

File: scripts/routing.py
"""
Budget-aware routing logic for the Web Doc Resolver.
"""

from dataclasses import dataclass
from urllib.parse import urlparse

from scripts.routing_memory import RoutingMemory


@dataclass
class ResolutionBudget:
    max_provider_attempts: int
    max_paid_attempts: int
    max_total_latency_ms: int
    allow_paid: bool = True
    quality_threshold: float = 0.65
    attempts: int = 0
    paid_attempts: int = 0
    elapsed_ms: int = 0
    stop_reason: str | None = None

    def can_try(self, *, is_paid: bool) -> bool:
        if self.attempts >= self.max_provider_attempts:
            self.stop_reason = "max_provider_attempts"
            return False
        if is_paid and not self.allow_paid:
            self.stop_reason = "paid_disabled"
            return False
        if is_paid and self.paid_attempts >= self.max_paid_attempts:
            self.stop_reason = "max_paid_attempts"
            return False
        if self.elapsed_ms >= self.max_total_latency_ms:
            self.stop_reason = "max_total_latency_ms"
            return False
        return True

    def record_attempt(self, *, is_paid: bool, latency_ms: int) -> None:
        self.attempts += 1
        self.elapsed_ms += latency_ms
        if is_paid:
            self.paid_attempts += 1


PROFILE_BUDGETS = {
    "free": {
        "max_provider_attempts": 3,
        "max_paid_attempts": 0,
        "max_total_latency_ms": 6000,
        "allow_paid": False,
        "quality_threshold": 0.70,
    },
    "balanced": {
        "max_provider_attempts": 6,
        "max_paid_attempts": 2,
        "max_total_latency_ms": 12000,
        "allow_paid": True,
        "quality_threshold": 0.65,
    },
    "fast": {
        "max_provider_attempts": 2,
        "max_paid_attempts": 1,
        "max_total_latency_ms": 4000,
        "allow_paid": True,
        "quality_threshold": 0.60,
    },
    "quality": {
        "max_provider_attempts": 10,
        "max_paid_attempts": 5,
        "max_total_latency_ms": 20000,
        "allow_paid": True,
        "quality_threshold": 0.55,
    },
}


def extract_domain(url: str) -> str | None:
    try:
        parsed = urlparse(url)
        return parsed.netloc.lower() or None
    except Exception:
        return None


def detect_doc_platform(url: str) -> str | None:
    """Detect common documentation platforms from URL."""
    try:
        parsed = urlparse(url)
    except Exception:
        return None

    hostname = (parsed.hostname or "").lower()
    path = (parsed.path or "").lower()

    if hostname == "gitbook.io" or hostname.endswith(".gitbook.io"):
        return "gitbook"
    if hostname == "gitbook.com" or hostname.endswith(".gitbook.com"):
        return "gitbook"
    if hostname == "readthedocs.io" or hostname.endswith(".readthedocs.io"):
        return "sphinx"
    if hostname == "rtfd.io" or hostname.endswith(".rtfd.io"):
        return "sphinx"
    if hostname == "www.mkdocs.org" or hostname == "mkdocs.org":
        return "mkdocs"
    if hostname == "notion.so" or hostname.endswith(".notion.so"):
        return "notion"
    if hostname == "notion.site" or hostname.endswith(".notion.site"):
        return "notion"
    if (
        (hostname.endswith(".atlassian.net") and path.startswith("/wiki"))
        or "confluence" in hostname
        or "confluence" in path
    ):
        return "confluence"

    return None


def preflight_route(url: str) -> dict:
    """
    Cheap preflight classifier: inspect URL signals to route to best extraction strategy.

    Returns dict with:
    - platform: detected doc platform or None
    - preferred_strategy: 'direct_fetch', 'llms_txt', 'extraction', 'browser'
    - confidence: 0.0-1.0
    - js_heavy: bool - likely needs JS rendering
    """
    platform = detect_doc_platform(url)
    parsed = urlparse(url)
    hostname = (parsed.hostname or "").lower()
    path = (parsed.path or "").lower()

    # Static doc platforms: prefer llms.txt → direct fetch
    if platform in ("gitbook", "sphinx", "mkdocs"):
        return {
            "platform": platform,
            "preferred_strategy": "llms_txt",
            "confidence": 0.85,
            "js_heavy": False,
        }

    # JS-heavy platforms: need browser or extraction
    if platform in ("notion", "confluence"):
        return {
            "platform": platform,
            "preferred_strategy": "extraction",
            "confidence": 0.8,
            "js_heavy": True,
        }

    # Known doc sites
    doc_signals = ["docs.", "doc.", "documentation", "/docs/", "/doc/", "/api/", "/reference/"]
    if any(s in hostname or s in path for s in doc_signals):
        return {
            "platform": None,
            "preferred_strategy": "llms_txt",
            "confidence": 0.6,
            "js_heavy": False,
        }

    # GitHub/GitLab: direct content
    if any(d in hostname for d in ["github.com", "gitlab.com", "bitbucket.org"]):
        return {
            "platform": None,
            "preferred_strategy": "direct_fetch",
            "confidence": 0.7,
            "js_heavy": False,
        }

    # Default: try llms.txt first, then extraction
    return {
        "platform": None,
        "preferred_strategy": "llms_txt",
        "confidence": 0.4,
        "js_heavy": False,
    }


def plan_provider_order(
    *,
    target: str,
    is_url: bool,
    custom_order: list[str] | None = None,
    skip_providers: set[str] | None = None,
    routing_memory: RoutingMemory | None = None,
) -> list[str]:
    if custom_order:
        base = list(custom_order)
    elif is_url:
        preflight = preflight_route(target)
        platform = preflight.get("platform")
        strategy = preflight.get("preferred_strategy", "llms_txt")

        if platform in ("notion", "confluence") or preflight.get("js_heavy"):
            base = ["firecrawl", "mistral_browser", "jina", "direct_fetch", "duckduckgo"]
        elif strategy == "direct_fetch":
            base = [
                "direct_fetch",
                "llms_txt",
                "jina",
                "firecrawl",
                "mistral_browser",
                "duckduckgo",
            ]
        else:
            base = [
                "llms_txt",
                "jina",
                "firecrawl",
                "direct_fetch",
                "mistral_browser",
                "duckduckgo",
            ]
    else:
        # DuckDuckGo deprioritized due to instability (Alert 2026-04-20)
        base = ["exa_mcp", "exa", "tavily", "serper", "mistral_websearch", "duckduckgo"]

    skip_providers = skip_providers or set()

    domain = extract_domain(target) if is_url else "query"
    if domain and routing_memory:
        base = routing_memory.rank(domain, base)

    return [p for p in base if p not in skip_providers]
